p2panda-net/src/sync/manager.rs

```diff
@@ -68,86 +177,90 @@ impl SyncManager {
                 .or_default()
                 .contains(&peer);
 
-            // Attempt connection in order to initiate a sync session.
+            // Schedule a sync attempt.
             if !active_peers.contains(&peer) && !sync_complete {
-                active_peers.insert(peer);
-                // @TODO: We might want to disentangle this call from the `update_peer_topics`
-                // method so that it returns immediately and the `connect` method has its own life.
-                if let Err(err) = self.connect(peer, topic).await {
-                    warn!("outbound connection attempt failed: {}", err)
-                }
+                self.schedule_attempt(peer, topic).await?;
```

Comment:
If I see it correctly this ties the scheduler to the receipt of announcement messages. We could detach the logics a little bit more, by just learning about the topics here and then checking if we have free "sync slots", if yes, start an attempt. If not, that's fine.

If a sync session ended (`complete_successful_sync`) or all attempts failed plus the queue has some space for us (`complete_failed_sync` + `sync_attempt.attempts > self.max_retry_attempts` + `queue.len() < MAX_CONCURRENT_SYNC_SESSIONS`)  we can look into our topic table again and schedule a new session.

I think this a) unties the logic nicely from "announcements" b) prevents the queue to ever be full (see other comment somewhere else)

Reply:
Cool idea! Looking into this now.